xfrd.c

```diff
@@ -2739,6 +2741,70 @@ xfrd_handle_taskresult(xfrd_state_type* xfrd, struct task_list_d* task)
 		xfrd_process_zonestat_inc_task(xfrd, task);
 		break;
 #endif
+	case task_add_catzone: {
+		char* zname = (char*)task->zname;
+		char* pattern = zname + strlen(zname) + 1;
+		char* catname = pattern + strlen(pattern) + 1;
+		char* member_id = catname + strlen(catname) + 1;
+
+		struct zone_options* zopt = NULL;
+		pattern_options_type* patopt;
+
+		log_msg(LOG_INFO, "now adding catzone %s %s %s %s", zname, pattern, catname, member_id);
+
+		patopt = pattern_options_find(xfrd->nsd->options, pattern);
+		if (!patopt) {
+			patopt = pattern_options_create(xfrd->region);
+			patopt->pname = pattern;
+			pattern_options_add_modify(xfrd->nsd->options, patopt);
+		} 
+		zopt = zone_options_find(xfrd->nsd->options, dname_parse(xfrd->region, zname));
+		if (!zopt) {
+			zopt = zone_list_zone_insert(xfrd->nsd->options, zname, pattern, 0, 0);
```

Comment:
This adds it to the zone.list file, instead of `nsd_options_insert_zone` that just adds it to the list of zones in memory in the options structure. At position 0,0 in the file, would this mess up the zone.list bookkeeping, and is the catalog zone not separate from that.

Reply:
`options.h` has as comment for `zone_list_zone_insert`: `/* create zonelist entry, do not insert in file (called by _add) */` - should I change that as well?

Reply:
The `zone_list_add` function is used by `remote.c` where it adds the zone that needs to be configured additionally, and it adds it to the zonelist part of the config, in memory, and also in the zonelist part of the config on disk in the zonelist file. The zone structure then also needs to be added into the options in memory for the other processes in NSD, this is done by the `zone_list_zone_insert` function that is called from `difffile.c` when the task to edit that contents arrives from the tasklist. It would then not add it to the zonelist file, because one process is modifying the zonelist file, that is the xfrd process that does that. For the zone structure that is added in catalog zones, I am not sure what it is that needs to be done, hence I was asking as well, perhaps it needs to add this in a different way. It should not modify the zonelist file on disk, and then also not change the 'zonelist file' data structure in memory?

Reply:
What I discussed with @wtoorop (in person) was that adding to the zone.list file itself is probably not necessary, as the catalog zone acts as kind of a zone list in and of itself. Let's see whether I can do that using `nsd_options_insert_zone` instead.